pkg/controller/statefulset/stateful_set_utils.go

```diff
@@ -132,6 +147,153 @@ func storageMatches(set *apps.StatefulSet, pod *v1.Pod) bool {
 	return true
 }
 
+// claimOwnerMatchesSetAndPod returns false if the ownerRefs of the claim are not set consistently with the
+// PVC deletion policy for the StatefulSet.
+func claimOwnerMatchesSetAndPod(claim *v1.PersistentVolumeClaim, set *apps.StatefulSet, pod *v1.Pod) bool {
+	switch set.Spec.PersistentVolumeClaimDeletePolicy {
+	case apps.Retain:
+		if hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnStatefulSetDeletionOnly:
+		if !hasOwnerRef(claim, set) ||
+			hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownOnly:
+		if hasOwnerRef(claim, set) {
+			return false
+		}
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	case apps.DeleteOnScaledownAndStatefulSetDeletion:
+		podScaledDown := getOrdinal(pod) >= int(*set.Spec.Replicas)
+		// If a pod is scaled down, there should be no set ref and a pod ref;
+		// if the pod is not scaled down it's the other way around.
+		if podScaledDown == hasOwnerRef(claim, set) {
+			return false
+		}
+		if podScaledDown != hasOwnerRef(claim, pod) {
+			return false
+		}
+	default:
+		klog.Errorf("Unknown policy %v!", set.Spec.PersistentVolumeClaimDeletePolicy)
```

Comment:
should this return an error?

Reply:
This ftn only returns a bool. Since the policy should be validated, it seems excessive to return an error.

I have changed the behavior to match Retain, though, since I'm going to have to deal with a nil pointer since making the field optional in the API.